Add a card that has the element matching screenshots

harriswong opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/2174

#### Changes proposed in this Pull Request
- use `WCPaySettingsContext` context to get the UPE feature flag
- Display banner if feature flag is true, hide otherwise.

#### Follow-up
- ~Clicking "Get started" links to the onboarding wizard will be done in a follow up https://github.com/Automattic/woocommerce-payments/issues/2323~ (Changes are now included in this PR)

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions
1. Enable `_wcpay_feature_upe_settings_preview` and disable `_wcpay_feature_upe`  feature flag. Confirm seeing the banner:
![image](https://user-images.githubusercontent.com/572862/125879390-bf738c5c-79c1-44ee-8a0c-74cbe63743eb.png)

2. Click "Enable in your store". You should be redirected to `wp-admin/admin.php?page=wc-admin&task=woocommerce-payments--additional-payment-methods` 
![image](https://user-images.githubusercontent.com/572862/125879264-28dbe8f8-d0b9-4084-9244-4dc14320ca71.png)

3. Check database `SELECT * FROM `wp_options` where option_name like '%_wcpay_feature_upe%'`. You should see there is an entry and the value is set to `1`. 
4. Disable `_wcpay_feature_upe_settings_preview` feature flag. Confirm banner is hidden:
![image](https://user-images.githubusercontent.com/572862/123486264-ecec0b80-d5c8-11eb-9831-b81ec32783fc.png)

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

waclawjacek commented on client/payment-methods/test/index.js:
```diff
@@ -143,4 +144,40 @@ describe( 'PaymentMethods', () => {
+		expect( enableWooCommercePaymentText ).toBeInTheDocument();
+	} );
+
+	test( 'express payments should not rendered when UPE preview feture flag is disabled', () => {
+		const featureFlagContext = {
+			featureFlags: { upeSettingsPreview: false, upe: false },
```
How about also adding a test for the `upe = 1, upe preview = 1` ("UPE available but already enabled") scenario, and maybe also `upe = 1, upe preview = 0`?

You could refactor the second test that's already in place to use [`test.each()`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) for this:

```suggestion
	test.each( [ [ false, false ], [ false, true ], [ true, true ] ] )( 'express payments should not rendered when UPE preview = %s and UPE = %s', ( upeSettingsPreview, upe ) => {
		const featureFlagContext = {
			featureFlags: { upeSettingsPreview, upe },
```

harriswong replied:
I like the idea of using a data provider to test the combinations and their results! Thanks!

waclawjacek reviewed (changes requested):
Great PR! 🎉 Works exactly as described.

One thing that seems to require tweaking is the padding so the text doesn't overlap the background image.

harriswong commented on client/payment-methods/style.scss:
```diff
@@ -27,4 +27,27 @@
 		}
 	}
+
+	&__express-checkouts {
```
Do you mean to add padding on the right side of the text so that the text wraps to another line in normal view. If it's in mobile view, then we push the background image out from view port, leaving just the purple background?

harriswong replied:
Updated so that the text no longer hover on top of the background image:
![image](https://user-images.githubusercontent.com/572862/124981552-ef039080-dff2-11eb-926d-bc66b7404689.png)

![image](https://user-images.githubusercontent.com/572862/124981540-ec08a000-dff2-11eb-8c06-e22acdacfc32.png)

harriswong commented on client/payment-methods/index.js:
```diff
@@ -108,6 +124,44 @@ const PaymentMethods = () => {
+						<CardDivider />
+						<CardBody className="payment-methods__express-checkouts">
+							<span className="payment-methods__pill">
+								Early access
+							</span>
+							<h3>
+								Enable the new WooCommerce Payments checkout
+								experience
```
Nice catch on localization!

harriswong commented:
@waclawjacek I fixed the merge conflict with latest `develop`. I also merged the changes from https://github.com/Automattic/woocommerce-payments/pull/2402 into this PR. I updated the test instruction. Clicking on "Enable in your store" should now enable the UPE flag.

waclawjacek commented on client/payment-methods/test/index.js:
```diff
@@ -200,4 +244,43 @@ describe( 'PaymentMethods', () => {
 		expect( disableUPEButton ).not.toBeInTheDocument();
 	} );
+
+	test( 'clicking "Enable in your store" in express payments enable UPE and redirects', () => {
```
You could do this to test if the redirect happens:

```diff
diff --git a/client/payment-methods/test/index.js b/client/payment-methods/test/index.js
index 855908f2..e48c7778 100644
--- a/client/payment-methods/test/index.js
+++ b/client/payment-methods/test/index.js
@@ -245,12 +245,10 @@ describe( 'PaymentMethods', () => {
 		expect( disableUPEButton ).not.toBeInTheDocument();
 	} );
 
-	test( 'clicking "Enable in your store" in express payments enable UPE and redirects', () => {
+	test( 'clicking "Enable in your store" in express payments enable UPE and redirects', async () => {
 		Object.defineProperty( window, 'location', {
-			get() {
-				return {
-					href: 'example.com/',
-				};
+			value: {
+				href: 'example.com/',
 			},
 		} );
 
@@ -280,7 +278,8 @@ describe( 'PaymentMethods', () => {
 		expect( enableInYourStoreButton ).toBeInTheDocument();
 
 		expect( setIsUpeEnabledMock ).not.toHaveBeenCalled();
-		user.click( enableInYourStoreButton );
+		await user.click( enableInYourStoreButton );
 		expect( setIsUpeEnabledMock ).toHaveBeenCalledWith( true );
+		expect( window.location.href ).toEqual( 'admin.php?page=wc-admin&task=woocommerce-payments--additional-payment-methods' );
 	} );
 } );
```

harriswong replied:
Thanks! Added this to verify href.

waclawjacek commented on client/payment-methods/index.js:
```diff
@@ -124,6 +112,56 @@ const UpeDisableError = () => {
+
+	return (
+		<>
+			<CardDivider />
+			<CardBody className="payment-methods__express-checkouts">
```
This now happens. I think it's due to the way the components are conditionally displayed on the settings page.

![Markup on 2021-07-16 at 07:53:55](https://user-images.githubusercontent.com/1759681/125904376-730b98e9-08e1-4e7e-a419-44533804adfd.png)

harriswong replied:
I think this issue is being address in these 2 places separately: 
https://github.com/Automattic/woocommerce-payments/pull/2423
https://github.com/Automattic/woocommerce-payments/pull/2456

frosso replied:
That's correct - the section should display the "card" payment method when UPE is disabled. Those 2 PRs will introduce the correct behavior.

waclawjacek replied:
Oh, I see! Thanks for pointing it out!

waclawjacek reviewed (changes requested):
Still a great PR but requesting changes again 😬 due to the UI quirks. I've attached screenshots where it made sense.

waclawjacek commented:
Oh, also:

![Markup on 2021-07-16 at 13:06:29](https://user-images.githubusercontent.com/1759681/125938551-93ac8f10-7c63-4a37-999b-8f91f51b777d.png)

This doesn't require a fix but if you'd like to know, there is one breakpoint that could apply a bit faster or something. I chose the least photogenic possible screen width for this screenshot. 😄 On Firefox, this was `965 x 912`.

harriswong commented:
> This doesn't require a fix but if you'd like to know, there is one breakpoint that could apply a bit faster or something. I chose the most buggy screen width I could find for this screenshot so as you can see, it's pretty good anyway. 😄 On Firefox, this was 965 x 912.

@waclawjacek Do you mean the background image should not show < 912 px width instead of the 600px it is at?

harriswong commented on client/payment-methods/index.js:
```diff
@@ -124,6 +112,56 @@ const UpeDisableError = () => {
+				</h3>
+				<p>
+					{ __(
+						'Get early access to additional payment methods and an improved checkout experience, coming soon to WooCommerce Payments.',
```
Is there a reason why the existing files follow the pattern of have this line at the top of file? Is it preferred to do it on a line-per-line basis or should it be at the top of file?

![image](https://user-images.githubusercontent.com/572862/125967771-4fbb9600-5c98-414b-85cd-8d64053c0a0c.png)

waclawjacek replied:
Using `/* eslint-disable max-len */` disables ESLint's `max-len` rule from that point onward. You can then re-enable it using `/* eslint-enable max-len */`. Placing it at the top of the file disables the rule for the entire file.

The "next line" approach is more strict in the sense that it only disables the rule for the next line so the rest of the file is still checked, just not this line.

I've noticed that some of the files are JSified SVGs so it makes sense to disable it for these.

harriswong replied:
Updated.

harriswong commented on client/payment-methods/index.js:
```diff
@@ -164,7 +209,7 @@ const PaymentMethods = () => {
 				) }
 				<CardBody>
-					{ isUpeEnabled ? (
+					{ isUpeEnabled && (
```
According to our design, the "credit card/debit card" section should actually be present throughout:
![image](https://user-images.githubusercontent.com/572862/125968345-a2206b74-4d34-48cd-adba-4f0934fa639d.png)

Clicking the button should not hide the top section.

frosso replied:
That's correct 👍

waclawjacek replied:
Awesome!

waclawjacek commented:
> @waclawjacek Do you mean the background image should not show < 912 px width instead of the 600px it is at?

Mmm, I'm not sure, it looks alright. I just wanted to point out there are some situations where the text overlaps the background -- not saying this should change though.

harriswong commented:
Ok i will double check with design and see if we need to adjust it. Thanks!

harriswong commented:
> I just wanted to point out there are some situations where the text overlaps the background

@waclawjacek I changed it to 1024 pixels. Tested in firefox. Text doesn't overlap:
![image](https://user-images.githubusercontent.com/572862/126190304-4f89f642-9699-4797-b1b5-edc600214cfc.png)

waclawjacek reviewed (approved):
Thank you for including all the changes! LGTM! :shipit:

waclawjacek commented on client/payment-methods/index.js:
```diff
@@ -128,9 +128,7 @@ const UpeSetupBanner = () => {
-				<span className="payment-methods__pill">
-					{ __( 'Early access', 'woocommerce-payments' ) }
-				</span>
+				<Pill>Early access</Pill>
```
```suggestion
				<Pill>{ __( 'Early access', 'woocommerce-payments' ) }</Pill>
```

harriswong replied:
D'oh. Thanks!

harriswong replied:
Also fixed the other `Early access` in this file.

waclawjacek reviewed (approved):
Just one tiny tweak and it's good to go!
